language-server/json_serialize.zig

```diff
@@ -0,0 +1,405 @@
+const std = @import("std");
+
+const debug = std.debug;
+const mem = std.mem;
+const heap = std.heap;
+const json = std.json;
+const expect = std.testing.expect;
+
+pub fn MaybeDefined(comptime T: type) type {
+    return union(enum) {
+        NotDefined,
+        Defined: T,
+
+        const __maybe_defined = void;
+    };
+}
+
+fn getMaybeDefinedChildType(comptime T: type) ?type {
+    switch (@typeInfo(T)) {
+        .Union => |unionInfo| {
+            if (unionInfo.decls.len == 1 and comptime mem.eql(u8, unionInfo.decls[0].name, "__maybe_defined")) {
+                return unionInfo.fields[1].field_type;
+            } else {
+                return null;
+            }
+        },
+        else => return null,
+    }
+}
+
+pub const Structured = union(enum) {
+    Array: json.Array,
+    Object: json.ObjectMap,
+};
+
+pub const Primitive = union(enum) {
+    Bool: bool,
+    Integer: i64,
+    Float: f64,
+    String: []const u8,
+};
+
+pub const Number = union(enum) {
+    Integer: i64,
+    Float: f64,
+};
+
+/// jsonStream must be a pointer to std.json.WriteStream
+pub fn serialize(value: var, jsonStream: var) !void {
```

Comment:
I have a WIP branch for new json stuff over at https://github.com/ziglang/zig/compare/master...daurnimator:auto-json
Will hopefully reopen the PR for it soon.

Reply:
Your branch looks nice, but I need the capabilities to not emit a struct field at all and tell the difference between `null` and not defined (see my `MaybeDefined`).

Reply:
1. use a union of a `void` and the type?
2. use the custom dumper?

Reply:
- I don't want to write a custom dumper for every struct that has a field that may not be defined.
- I need the parser to distinguish between null and not defined.

Reply:
> I need the parser to distinguish between null and not defined.

It does: not defined values use their default values.